macros/src/main/scala/org/zalando/grafter/macros/ReaderMacros.scala

```diff
@@ -0,0 +1,91 @@
+package org.zalando.grafter.macros
+
+import scala.reflect.macros.whitebox
+
+object ReaderMacros {
+
+  /** get the annotated class and, if available, companion object */
+  def annotationInputs(c: whitebox.Context)(name: String)(annottees: Seq[c.Expr[Any]]): (c.universe.Tree, Option[c.universe.Tree]) =
+    annottees.toList match {
+      case classDecl :: companion :: _ =>
+      (classDecl.tree, Option(companion.tree))
+
+      case classDecl :: _ =>
+      (classDecl.tree, None)
+
+      case Nil => c.abort(c.enclosingPosition, s"the @$name annotation must annotate a class")
+    }
+
+  /**
+   * Inject an expression in the companion object of a class.
+   *
+   * Create one if necessary
+   */
+  def outputs(c: whitebox.Context)(
+    classTree: c.universe.Tree,
+    className: c.universe.TypeName,
+    companionTree: Option[c.universe.Tree])(expression: c.universe.Tree) = {
+
+    import c.universe._
+
+    val insert = c.Expr[Any](expression)
+    val out: List[Tree] = {
+      val companionObject = companionTree match {
+        case Some(q"""$mod object $companionName extends { ..$earlydefns } with ..$parents { ..$body }""") =>
+          q"""$mod object $companionName extends { ..$earlydefns } with ..$parents {
+           ..$body
+           ..$insert
+           }"""
+
+        case None =>
+          q"""object ${TermName(className.decodedName.toString)} {
+           ..$insert
+           }"""
+      }
+
+      classTree :: companionObject :: Nil
+    }
+
+    c.Expr[Any](Block(out, Literal(Constant(()))))
```

Comment:
That's right, I took the previous version for granted. It can actually be simplified with a quasiquote. Thanks!